prow/git/v2/remote.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"net/url"
+	"path"
+)
+
+// RemoteResolverFactory knows how to construct remote resolvers for
+// authoritative central remotes (to pull from) and publish remotes
+// (to push to) for a repository. These resolvers
+type RemoteResolverFactory interface {
```

Comment:
To me, the names of `RemoteResolverFactory` methods do not correspond to what the implementations do: the only difference between the two methods is whether they make a resolver that uses the passed `org` param (`Central`) or constructs the "owner" (username or org) bit of the repo dynamically using the callbacks stored in implementations (`Publish`). But this does not make the resolved remotes "central" or "publishing". These terms seems to be defined more by the callsite (=defined by when and for what purpose are these called) than by what these methods actually do and offer to callers. To me this is very confusing.

Reply:
> But this does not make the resolved remotes "central" or "publishing".

Why not? The credentials we use to interact with GitHub are the ones that we can publish to, so the definition of a remote we publish to is the remote for which we have credentials, which are resolved at runtime since they are dynamic. The authoritative remote is static.

Reply:
And I guess "dynamic" and "static" describe much better what the two methods really do. The ability of publishing and being central is not handled by `remote.go` and `RemoteResolverFactory` at all, that's totally determined by whatever is using them.

Reply:
No, those are implementation details of how we create these things. If we did not re-load data from disk we would still have publish and authoritative remotes, but neither would be dynamic.